File: modules/openlmis-web/src/main/java/org/openlmis/web/controller/VaccineQuantificationController.java
/*
 * This program is part of the OpenLMIS logistics management information system platform software.
 * Copyright © 2013 VillageReach
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *  
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.
 * You should have received a copy of the GNU Affero General Public License along with this program.  If not, see http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org. 
 */

package org.openlmis.web.controller;

import org.openlmis.vaccine.domain.VaccineQuantification;
import org.openlmis.vaccine.service.VaccineQuantificationService;
import org.openlmis.web.response.OpenLmisResponse;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.dao.DuplicateKeyException;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;

import javax.servlet.http.HttpServletRequest;

import java.util.Date;

import static org.openlmis.web.response.OpenLmisResponse.success;
import static org.springframework.web.bind.annotation.RequestMethod.DELETE;
import static org.springframework.web.bind.annotation.RequestMethod.GET;
import static org.springframework.web.bind.annotation.RequestMethod.POST;

@Controller
@RequestMapping(value="/vaccineQuantification")
public class VaccineQuantificationController extends BaseController{


    @Autowired
    VaccineQuantificationService vaccineQuantificationService;

    @RequestMapping(value = "/create", method = POST, headers = ACCEPT_JSON)
    // TODO: Add appropriate permission
    //  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_VACCINE_QUANTIFICATION')")
    public ResponseEntity insert(@RequestBody VaccineQuantification vaccineQuantification, HttpServletRequest request) {

        vaccineQuantification.setCreatedBy(loggedInUserId(request));

        ResponseEntity<OpenLmisResponse> response;

        try {
            vaccineQuantificationService.updateVaccineQuantification(vaccineQuantification);
        } catch(DuplicateKeyException exp){
            return OpenLmisResponse.error("There is record with the same vaccine quantification year already.", HttpStatus.BAD_REQUEST);
        }

        response = success(messageService.message("message.vaccine.Quantification.created.success"));
        //response.getBody().addData("vaccineQuantification", vaccineQuantification);
        return response;
    }

    @RequestMapping(value = "/delete/{id}", method = DELETE, headers = ACCEPT_JSON)
    // TODO: Add appropriate permission
    //  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_VACCINE_QUANTIFICATION')")
    public ResponseEntity deleteVaccineQuantification(@PathVariable(value="id") Long id){

        vaccineQuantificationService.deleteVaccineQuantification(id);

        ResponseEntity<OpenLmisResponse> response;
        response = success(messageService.message("message.vaccine.Quantification.created.success"));
        return response;
    }

    @RequestMapping(value = "/list", method = GET, headers = ACCEPT_JSON)
    public ResponseEntity getVaccineQuantifications() {

        return OpenLmisResponse.response("vaccineQuantifications", vaccineQuantificationService.getVaccineQuantifications());
    }

    @RequestMapping(value = "/get/{id}", method = GET, headers = ACCEPT_JSON)
    public ResponseEntity getVaccineQuantification(@PathVariable(value="id") Long id){
        return OpenLmisResponse.response("vaccineQuantification", vaccineQuantificationService.getVaccineQuantification(id));
    }

    @RequestMapping(value = "/dilutionsList", method = GET, headers = ACCEPT_JSON)
    public ResponseEntity getVaccineQuantificationDilutionsList(@PathVariable(value="id") Long id){
        return OpenLmisResponse.response("dilution", vaccineQuantificationService.getVaccineDilutions());
    }

    @RequestMapping(value = "/administrationMode", method = GET, headers = ACCEPT_JSON)
    public ResponseEntity getVaccineQuantificationAdministrationMode(){
        return OpenLmisResponse.response("administrationMode", vaccineQuantificationService.getVaccineAdministrationMode());
    }


    @RequestMapping(value = "/vaccination_type", method = GET, headers = ACCEPT_JSON)
    public ResponseEntity getVaccineQuantificationVaccinationType(){
        return OpenLmisResponse.response("vaccinationType", vaccineQuantificationService.getVaccinationTypes());
    }

    @RequestMapping(value = "/formLookups", method = GET, headers = ACCEPT_JSON)
    public ResponseEntity getVaccineQuantificationFormLookps(){

        ResponseEntity<OpenLmisResponse> response;
        response = success("success");
        response.getBody().addData("dilution",  vaccineQuantificationService.getVaccineDilutions());
        response.getBody().addData("administrationMode", vaccineQuantificationService.getVaccineAdministrationMode());
        response.getBody().addData("vaccinationType", vaccineQuantificationService.getVaccinationTypes());

        return response;
    }



}

File: modules/openlmis-web/src/main/webapp/public/js/vaccine/controller/vaccine-quantification-controller.js
/*
 * This program was produced for the U.S. Agency for International Development. It was prepared by the USAID | DELIVER PROJECT, Task Order 4. It is part of a project which utilizes code originally licensed under the terms of the Mozilla Public License (MPL) v2 and therefore is licensed under MPL v2 or later.
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the Mozilla Public License as published by the Mozilla Foundation, either version 2 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the Mozilla Public License for more details.
 *
 * You should have received a copy of the Mozilla Public License along with this program. If not, see http://www.mozilla.org/MPL/
 */


function VaccineQuantificationController($scope, $dialog, messageService, $routeParams, $location, VaccineQuantificationFormLookUps, ReportProductsByProgram, VaccineQuantificationList, ProgramCompleteList, VaccineQuantificationUpdate, GetVaccineQuantification, DeleteVaccineQuantification) {

    $scope.vaccineQuantifications = {};


    if (isUndefined($routeParams.id) || $routeParams.id === 0) {

        VaccineQuantificationList.get({}, function(data){
            $scope.vaccineQuantifications = data.vaccineQuantifications;
        });

    } else {

        $scope.$parent.message = '';

        ProgramCompleteList.get({}, function(data){

            $scope.programs = data.programs;
        });

        VaccineQuantificationFormLookUps.get({}, function(data){
            $scope.administrationModes = data.administrationMode;
            $scope.dilutions = data.dilution;
            $scope.vaccinationTypes = data.vaccinationType;

            $scope.administrationModes.unshift({
                name : '--Select Administration mode--'
            });
            $scope.dilutions.unshift({
                name : '--Select Dilution --'
            });
            $scope.vaccinationTypes.unshift({
                name : '--Select Vaccination Type --'
            });
        });

        GetVaccineQuantification.get({
            id: $routeParams.id
        }, function (data) {
            $scope.quantification = data.vaccineQuantification;
        });
    }

    $scope.saveVaccineQuantification = function(){
        if ($scope.vaccineQuantificationForm.$error.pattern || $scope.vaccineQuantificationForm.$error.min || $scope.vaccineQuantificationForm.$error.required) {
            $scope.showError = "true";
            $scope.error = 'form.error';
            $scope.message = "";
            return;
        }

        var onSuccess = function(data){
            $scope.$parent.message = 'Your changes have been saved!';
            loadVaccineQuantificationList();
            $location.path('/quantification');
        };

        var onError = function(data){
            $scope.showError = true;
            $scope.error = data.data.error;
        };

        VaccineQuantificationUpdate.post($scope.quantification,onSuccess, onError);
    };

    $scope.cancelQuantificationForm = function(){
        $location.path('/quantification');
    };


    function loadVaccineQuantificationList(){
        VaccineQuantificationList.get({}, function(data){
            $scope.vaccineQuantifications = data.vaccineQuantifications;
        });
    }

    $scope.showRemoveVaccineQuantificationDialog = function (index) {
        var vaccineQuantification = $scope.vaccineQuantifications[index];
        $scope.selectedVaccineQuantification = vaccineQuantification;

        var options = {
            id: "removeVaccineQuantificationConfirmDialog",
            header: "Confirmation",
            body: "Please confirm that you want to delete the selected vaccine Quantification"
        };
        OpenLmisDialog.newDialog(options, $scope.removeVaccineQuantificationConfirm, $dialog, messageService);
    };

    $scope.removeVaccineQuantificationConfirm = function (result) {

        if (result) {
            DeleteVaccineQuantification.delete({'id': $scope.selectedVaccineQuantification.id}, function(){
                loadVaccineQuantificationList();
                $scope.$parent.message = 'Vaccine Quantification deleted successfully!';
            });
        }
    };

    $scope.$watch('quantification.programId', function (value) {

        if(isUndefined(value)){
            $scope.products = {};
        }
        else{
            ReportProductsByProgram.get({
                programId: value
            }, function (data) {
                $scope.products = data.productList;
            });
        }

    });

}